## London Crime Data Analysis

#### In this notebook, I used pyspark for big data analysis for London crime data. I downloaded the data from https://data.police.uk/data/ between Dec 2018 and Dec 2020 which was provided by Metropolitan Police Service

In [0]:
from pyspark.sql import Window
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *

import os
os.environ["PYSPARK_PYTHON"] = "python3"

In [0]:
spark = SparkSession \
    .builder \
    .appName("crime analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

In [0]:
display(dbutils.fs.ls("dbfs:/FileStore/london_crime/"))

path name size dbfs:/FileStore/london_crime/london_201812.csv london_201812.csv 20464497 dbfs:/FileStore/london_crime/london_201901.csv london_201901.csv 20727597 dbfs:/FileStore/london_crime/london_201902.csv london_201902.csv 20105097 dbfs:/FileStore/london_crime/london_201903.csv london_201903.csv 22250428 dbfs:/FileStore/london_crime/london_201904.csv london_201904.csv 20889276 dbfs:/FileStore/london_crime/london_201905.csv london_201905.csv 21867296 dbfs:/FileStore/london_crime/london_201906.csv london_201906.csv 21828155 dbfs:/FileStore/london_crime/london_201907.csv london_201907.csv 22378157 dbfs:/FileStore/london_crime/london_201908.csv london_201908.csv 20753493 dbfs:/FileStore/london_crime/london_201909.csv london_201909.csv 20120458 dbfs:/FileStore/london_crime/london_201910.csv london_201910.csv 22409235 dbfs:/FileStore/london_crime/london_201911.csv london_201911.csv 21327099 dbfs:/FileStore/london_crime/london_201912.csv london_201912.csv 20189823 dbfs:/FileStore/london_crime/london_202001.csv london_202001.csv 21358026 dbfs:/FileStore/london_crime/london_202002.csv london_202002.csv 20343198 dbfs:/FileStore/london_crime/london_202003.csv london_202003.csv 20459832 dbfs:/FileStore/london_crime/london_202004.csv london_202004.csv 22284892 dbfs:/FileStore/london_crime/london_202005.csv london_202005.csv 23649395 dbfs:/FileStore/london_crime/london_202006.csv london_202006.csv 21637000 dbfs:/FileStore/london_crime/london_202007.csv london_202007.csv 23438533 dbfs:/FileStore/london_crime/london_202008.csv london_202008.csv 23709039 dbfs:/FileStore/london_crime/london_202009.csv london_202009.csv 22753146 dbfs:/FileStore/london_crime/london_202010.csv london_202010.csv 22538282 dbfs:/FileStore/london_crime/london_202011.csv london_202011.csv 21268730

In [0]:
df = spark.read.format('csv').option('header', 'true').load('dbfs:/FileStore/london_crime/*.csv')

In [0]:
print(df.count(), len(df.columns))

2289345 12

In [0]:
display(df)

Crime ID Month Reported by Falls within Longitude Latitude Location LSOA code LSOA name Crime type Last outcome category Context 4704f5ef984cf84ac5f1de28c7ef83dc7c98b3626edd9ad30c609b0e363e6f3c 2020-08 Metropolitan Police Service Metropolitan Police Service -0.744130 50.770195 On or near Martlet Way E01031448 Arun 019E Other theft Under investigation null 24771cd920c6e47f6e5e8b8aef790323a7e21f4d334f37f502ef2a2329ce0c62 2020-08 Metropolitan Police Service Metropolitan Police Service 0.834571 51.143435 On or near Coronation Drive E01032814 Ashford 012F Violence and sexual offences Under investigation null null 2020-08 Metropolitan Police Service Metropolitan Police Service 0.145888 51.593835 On or near Providence Place E01000027 Barking and Dagenham 001A Anti-social behaviour null null null 2020-08 Metropolitan Police Service Metropolitan Police Service 0.137065 51.583672 On or near Police Station E01000027 Barking and Dagenham 001A Anti-social behaviour null null null 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140035 51.589112 On or near Beansland Grove E01000027 Barking and Dagenham 001A Anti-social behaviour null null null 2020-08 Metropolitan Police Service Metropolitan Police Service 0.134947 51.588063 On or near Mead Grove E01000027 Barking and Dagenham 001A Anti-social behaviour null null 722559ad01cb40830f39483f5e947f8c93b3a6f54058466ddf6cf2ab6ce5d9c7 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140192 51.582311 On or near Hatch Grove E01000027 Barking and Dagenham 001A Criminal damage and arson Awaiting court outcome null 9b4e320c7a6dc6538a2f51b846306d200672b9b60cbfc1b8e5572bf7d9e88b24 2020-08 Metropolitan Police Service Metropolitan Police Service 0.137065 51.583672 On or near Police Station E01000027 Barking and Dagenham 001A Criminal damage and arson Investigation complete; no suspect identified null 318b49f3f96f5b2c250470e342ce4d203c27f64652b7adf90a2f7de9556efcaa 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140192 51.582311 On or near Hatch Grove E01000027 Barking and Dagenham 001A Criminal damage and arson Awaiting court outcome null 67b5250077bbc52ca154ed9f1a2685fb35630a506d9b9a486fe078dbd427c9c3 2020-08 Metropolitan Police Service Metropolitan Police Service 0.135866 51.587336 On or near Gibbfield Close E01000027 Barking and Dagenham 001A Criminal damage and arson Awaiting court outcome null 722559ad01cb40830f39483f5e947f8c93b3a6f54058466ddf6cf2ab6ce5d9c7 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140192 51.582311 On or near Hatch Grove E01000027 Barking and Dagenham 001A Criminal damage and arson Under investigation null daa331a949680eb163100381eeb90bde15f33196ea7f6bdd7aa1c6d11ceb807d 2020-08 Metropolitan Police Service Metropolitan Police Service 0.136513 51.588214 On or near Kingston Close E01000027 Barking and Dagenham 001A Criminal damage and arson Investigation complete; no suspect identified null 83377cc66c6315eea2fcc95f706017e55e9620cfc89926e0d4491b24e797fc03 2020-08 Metropolitan Police Service Metropolitan Police Service 0.135866 51.587336 On or near Gibbfield Close E01000027 Barking and Dagenham 001A Drugs Awaiting court outcome null fba5cae84bccf8c2a26fd7b6101c91d1434ee5817b9454a9b1d2e4fd10a8316d 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140634 51.583427 On or near Rams Grove E01000027 Barking and Dagenham 001A Other theft Investigation complete; no suspect identified null c6aa373d40bb89281ed2e1cc51be940d6380ebb94a987f16080a6722123711f8 2020-08 Metropolitan Police Service Metropolitan Police Service 0.148434 51.595164 On or near Park/Open Space E01000027 Barking and Dagenham 001A Public order Investigation complete; no suspect identified null 0c04429e6c3422a036beaa4cc82f7fdbf187dc2376f96178fed15a49c4e4f916 2020-08 Metropolitan Police Service Metropolitan Police Service 0.137065 51.583672 On or near Police Station E01000027 Barking and Dagenham 001A Public order Under investigation null b8c4ca

In [0]:
df = df.withColumnRenamed('LSOA code', 'LSOACode')\
  .withColumnRenamed('LSOA name', 'LSOAName')\
  .withColumnRenamed('Crime type', 'CrimeType')\
  .withColumnRenamed('Last outcome category', 'LastOutcomeCategory')

display(df)

Crime ID Month Reported by Falls within Longitude Latitude Location LSOACode LSOAName CrimeType LastOutcomeCategory Context 4704f5ef984cf84ac5f1de28c7ef83dc7c98b3626edd9ad30c609b0e363e6f3c 2020-08 Metropolitan Police Service Metropolitan Police Service -0.744130 50.770195 On or near Martlet Way E01031448 Arun 019E Other theft Under investigation null 24771cd920c6e47f6e5e8b8aef790323a7e21f4d334f37f502ef2a2329ce0c62 2020-08 Metropolitan Police Service Metropolitan Police Service 0.834571 51.143435 On or near Coronation Drive E01032814 Ashford 012F Violence and sexual offences Under investigation null null 2020-08 Metropolitan Police Service Metropolitan Police Service 0.145888 51.593835 On or near Providence Place E01000027 Barking and Dagenham 001A Anti-social behaviour null null null 2020-08 Metropolitan Police Service Metropolitan Police Service 0.137065 51.583672 On or near Police Station E01000027 Barking and Dagenham 001A Anti-social behaviour null null null 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140035 51.589112 On or near Beansland Grove E01000027 Barking and Dagenham 001A Anti-social behaviour null null null 2020-08 Metropolitan Police Service Metropolitan Police Service 0.134947 51.588063 On or near Mead Grove E01000027 Barking and Dagenham 001A Anti-social behaviour null null 722559ad01cb40830f39483f5e947f8c93b3a6f54058466ddf6cf2ab6ce5d9c7 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140192 51.582311 On or near Hatch Grove E01000027 Barking and Dagenham 001A Criminal damage and arson Awaiting court outcome null 9b4e320c7a6dc6538a2f51b846306d200672b9b60cbfc1b8e5572bf7d9e88b24 2020-08 Metropolitan Police Service Metropolitan Police Service 0.137065 51.583672 On or near Police Station E01000027 Barking and Dagenham 001A Criminal damage and arson Investigation complete; no suspect identified null 318b49f3f96f5b2c250470e342ce4d203c27f64652b7adf90a2f7de9556efcaa 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140192 51.582311 On or near Hatch Grove E01000027 Barking and Dagenham 001A Criminal damage and arson Awaiting court outcome null 67b5250077bbc52ca154ed9f1a2685fb35630a506d9b9a486fe078dbd427c9c3 2020-08 Metropolitan Police Service Metropolitan Police Service 0.135866 51.587336 On or near Gibbfield Close E01000027 Barking and Dagenham 001A Criminal damage and arson Awaiting court outcome null 722559ad01cb40830f39483f5e947f8c93b3a6f54058466ddf6cf2ab6ce5d9c7 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140192 51.582311 On or near Hatch Grove E01000027 Barking and Dagenham 001A Criminal damage and arson Under investigation null daa331a949680eb163100381eeb90bde15f33196ea7f6bdd7aa1c6d11ceb807d 2020-08 Metropolitan Police Service Metropolitan Police Service 0.136513 51.588214 On or near Kingston Close E01000027 Barking and Dagenham 001A Criminal damage and arson Investigation complete; no suspect identified null 83377cc66c6315eea2fcc95f706017e55e9620cfc89926e0d4491b24e797fc03 2020-08 Metropolitan Police Service Metropolitan Police Service 0.135866 51.587336 On or near Gibbfield Close E01000027 Barking and Dagenham 001A Drugs Awaiting court outcome null fba5cae84bccf8c2a26fd7b6101c91d1434ee5817b9454a9b1d2e4fd10a8316d 2020-08 Metropolitan Police Service Metropolitan Police Service 0.140634 51.583427 On or near Rams Grove E01000027 Barking and Dagenham 001A Other theft Investigation complete; no suspect identified null c6aa373d40bb89281ed2e1cc51be940d6380ebb94a987f16080a6722123711f8 2020-08 Metropolitan Police Service Metropolitan Police Service 0.148434 51.595164 On or near Park/Open Space E01000027 Barking and Dagenham 001A Public order Investigation complete; no suspect identified null 0c04429e6c3422a036beaa4cc82f7fdbf187dc2376f96178fed15a49c4e4f916 2020-08 Metropolitan Police Service Metropolitan Police Service 0.137065 51.583672 On or near Police Station E01000027 Barking and Dagenham 001A Public order Under investigation null b8c4ca2675a

In [0]:
# Drop duplicates
df = df.dropDuplicates()
print(df.count())

1922583

In [0]:
for c in df.columns:
  print(df.agg(countDistinct(c)).collect())        

[Row(count(Crime ID)=1560719)]
[Row(count(Month)=24)]
[Row(count(Reported by)=1)]
[Row(count(Falls within)=1)]
[Row(count(Longitude)=60848)]
[Row(count(Latitude)=57883)]
[Row(count(Location)=36897)]
[Row(count(LSOACode)=7509)]
[Row(count(LSOAName)=7509)]
[Row(count(CrimeType)=14)]
[Row(count(LastOutcomeCategory)=24)]
[Row(count(Context)=0)]

* 24 distinct months from Dec 2018 to Dec 2020
* All crimes are reported by Metropolitan Police Service
* All crimes fall within Metropolitan Police Service
* There are 7509 LSOAName, each LSOAName has a LSOANameCode. Lower Layer Super Output Areas (LSOA) are a geographic hierarchy designed to improve the reporting of small area statistics in England and Wales.
* There are 14 types of crime in this dataset
* There are 24 types of outcome of the crime in this dataset

In [0]:
df.createOrReplaceTempView('londonCrime')

## Q1. How does the number of crimes change by month?

Number of crimes dipped significantly in Apr 2020, because it was the first COVID19 lockdown. Then we had a second, a third, a fourth... lockdown. Criminals are not bothered anymore.

In [0]:
%sql
select Month as month_of_the_year, count(*) as number_of_crimes
from londonCrime
group by 1
order by 1

month_of_the_year number_of_crimes 2018-12 79099 2019-01 80190 2019-02 77731 2019-03 86149 2019-04 81044 2019-05 84684 2019-06 83700 2019-07 84889 2019-08 79007 2019-09 77273 2019-10 85850 2019-11 81715 2019-12 77644 2020-01 82042 2020-02 78448 2020-03 75782 2020-04 68500 2020-05 75775 2020-06 75674 2020-07 83362 2020-08 84012 2020-09 82137 2020-10 81380 2020-11 76496

### Q2: What are the top 10 crime locations?

On or near supermarket, parking area, shopping area, petrol station, sports/recreation area, nightclub.....

In [0]:
%sql
select Location, count(*) as number_of_crimes
from londonCrime
group by 1
order by number_of_crimes desc

Location number_of_crimes On or near Supermarket 70218 On or near Parking Area 50751 On or near Shopping Area 47283 On or near Petrol Station 35169 On or near Sports/Recreation Area 25468 On or near Nightclub 23978 No Location 18966 On or near Pedestrian Subway 17489 On or near Further/Higher Educational Building 13414 On or near Theatre/Concert Hall 12951 On or near Hospital 11596 On or near Police Station 9637 On or near High Street 5592 On or near Conference/Exhibition Centre 5435 On or near Park/Open Space 3739 On or near Bus/Coach Station 3665 On or near Station Road 3264 On or near High Road 2810 On or near London Road 2524 On or near Meridian Square 2510 On or near Station Approach 1746 On or near Church Road 1709 On or near Wallis Road 1663 On or near Airport/Airfield 1394 On or near Queen'S Road 1335 On or near Park Road 1332 On or near Grove Road 1311 On or near The Broadway 1270 On or near West Carriage Drive 1263 On or near Green Lanes 1194 On or near Uxbridge Road 1172 On or near Coventry Street 1100 On or near Manor Road 1073 On or near Victoria Road 1068 On or near Shaftesbury Avenue 1052 On or near Prison 1033 On or near Wood Lane 1031 On or near Myrtle Road 1030 On or near Westminster Bridge Road 1021 On or near The Green 1011 On or near Waterloo Road 995 On or near Holloway Road 986 On or near Stanley Road 942 On or near Avenue Road 934 On or near Charing Cross Road 929 On or near Coldharbour Lane 908 On or near York Road 890 On or near Lordship Lane 887 On or near Cranbrook Road 886 On or near Stroud Green Road 878 On or near Albert Road 876 On or near The Avenue 873 On or near King'S Road 861 On or near Forest Road 861 On or near Richmond Road 850 On or near Southwark Street 839 On or near Falconberg Mews 833 On or near Hertford Road 830 On or near Mare Street 822 On or near Theme/Adventure Park 815 On or near Salisbury Road 813 On or near Marlborough Road 811 On or near East Street 809 On or near The Drive 809 On or near Green Lane 802 On or near Borough High Street 791 On or near Warwick Road 790 On or near West Park Road 779 On or near Oxford Road 772 On or near Church Lane 756 On or near St John'S Road 754 On or near Essex Road 745 On or near Park Lane 742 On or near Cromwell Road 733 On or near Abbey Road 730 On or near Church Street 729 On or near Broadway 724 On or near Clifton Road 723 On or near Grosvenor Road 721 On or near Upper Street 720 On or near Ladbroke Grove 718 On or near Brighton Road 717 On or near Seven Sisters Road 712 On or near Kingsland Road 709 On or near Station Parade 703 On or near Gloucester Road 702 On or near Windsor Road 691 On or near Beaconsfield Road 686 On or near Moor Street 677 On or near Grange Road 675 On or near Harrow Road 667 On or near New Road 665 On or near Rivington Place 664 On or near Bedford Road 664 On or near Hamilton Road 652 On or near Cecil Road 651 On or near College Road 642 On or near Park Avenue 635 On or near Crescent Road 634 On or near Station Place 634 On or near Bath Road 634 On or near Buller Road 631 On or near Bromley Road 619 On or near Camden High Street 617 On or near North Road 614 On or near Roman Way 612 On or near Douglas Road 611 On or near Buckingham Mews 611 On or near Montague Road 608 On or near Lancaster Road 606 On or near Alexandra Road 602 On or near Gray'S Inn Road 599 On or near South Road 598 On or near Granville Road 596 On or near Little Newport Street 592 On or near Mile End Road 591 On or near Ferry Terminal 591 On or near Eresby Drive 591 On or near Whitcomb Street 590 On or near Finchley Road 590 On or near Vicarage Road 590 On or near Bridge Street 586 On or near Romford Road 586 On or near Barking Road 583 On or near Campbell Road 582 On or near Howard Road 581 On or near New Cross Road 579 On or near Cambridge Road 578 On or near Milton Road 576 On or near Bridge Road 576 On or near Victoria Street 573 On or near Harewood Place 570 On or near Silver Road 569 On or near Motorway Service Area 565 On or near James 

### Q3: What are the top crime Boroughs? 

* Top 32 are the 32 London Boroughs. Apart from 32 London Boroughs, Metropolitain Police also deals with small number of crimes in areas/cities outside London.
* **Top 5 London Boroughs by crime numbers are: Westminster, Southwark, Tower Hamlets, Newham and Camden. No surprise.**

In [0]:
%sql
select left(LSOAName,length(LSOAName) - 4) as Borough, count(*) as number_of_crimes
from londonCrime
group by 1
order by number_of_crimes desc

Borough number_of_crimes Westminster 136372 Southwark 78754 Tower Hamlets 76202 Newham 75519 Camden 75457 Lambeth 74240 Croydon 73791 Hackney 71012 Ealing 69384 Haringey 67003 Barnet 66772 Enfield 66479 Brent 65883 Hillingdon 64403 Lewisham 62902 Islington 62279 Greenwich 59367 Wandsworth 57445 Hounslow 57444 Waltham Forest 54647 Bromley 53615 Redbridge 53165 Hammersmith and Fulham 46469 Kensington and Chelsea 46455 Barking and Dagenham 42464 Havering 40953 Bexley 38782 Harrow 37816 Merton 32983 Sutton 31431 Richmond upon Thames 30122 Kingston upon Thames 27433 null 18966 City of London 1644 Epping Forest 382 Spelthorne 232 Dartford 186 Elmbridge 166 Epsom and Ewell 153 Thurrock 148 Broxbourne 141 Hertsmere 141 Reigate and Banstead 129 Tandridge 127 Slough 120 Medway 92 Three Rivers 85 Birmingham 79 Luton 78 South Bucks 77 Sevenoaks 76 Watford 69 Basildon 65 Brentwood 63 Welwyn Hatfield 57 Windsor and Maidenhead 48 Gravesham 45 Crawley 40 Dacorum 37 St Albans 36 Brighton and Hove 36 Manchester 33 Swale 31 Milton Keynes 31 Bradford 30 East Hertfordshire 28 Leicester 27 Harlow 27 Chelmsford 27 Central Bedfordshire 26 Tendring 26 Maidstone 26 Colchester 25 Bedford 25 Leeds 24 Tonbridge and Malling 24 Coventry 24 Thanet 23 Wycombe 23 Uttlesford 23 Runnymede 23 Reading 22 Cambridge 22 Castle Point 22 Chiltern 22 Northampton 22 Woking 19 Mole Valley 19 Liverpool 19 Cherwell 18 Mid Sussex 18 Bristol 18 Nottingham 18 Canterbury 17 Cardiff 17 Southampton 17 Peterborough 17 Southend-on-Sea 17 Wiltshire 16 Dover 16 Stevenage 16 Aylesbury Vale 15 St Edmundsbury 15 Shepway 14 County Durham 13 Horsham 13 Norwich 13 Bournemouth 13 Stoke-on-Trent 13 Forest Heath 13 Oldham 13 Great Yarmouth 13 Oxford 13 Wealden 12 Breckland 12 Arun 12 Swindon 12 Bolton 12 Braintree 12 Sheffield 12 Salford 12 Sandwell 12 Cheshire West and Chester 11 Ashford 11 Waveney 11 Ipswich 11 Hastings 11 West Berkshire 11 Bath and North East Somerset 11 Bracknell Forest 11 Huntingdonshire 11 Portsmouth 11 Kingston upon Hull 11 Guildford 10 Rushmoor 10 Cornwall 10 Kirklees 10 Newport 10 Chichester 10 New Forest 10 Lewes 9 Tunbridge Wells 9 Cheshire East 9 Rother 9 Derby 9 Worthing 9 Suffolk Coastal 9 Surrey Heath 9 Wolverhampton 9 Walsall 8 Wokingham 8 North Hertfordshire 8 Gloucester 8 Rochford 8 Wychavon 7 Hart 7 Wyre 7 King's Lynn and West Norfolk 7 Rochdale 7 Torbay 7 Warwick 7 Waverley 7 Tameside 7 Poole 6 Swansea 6 Rugby 6 Daventry 6 Basingstoke and Deane 6 Wakefield 6 Vale of White Horse 6 North Norfolk 6 Wirral 6 Forest of Dean 6 South Kesteven 6 Cheltenham 6 Mendip 6 South Oxfordshire 6 South Somerset 5 South Cambridgeshire 5 Rotherham 5 York 5 Sunderland 5 Wrexham 5 Gateshead 5 Middlesbrough 5 Cotswold 5 Dudley 5 Charnwood 5 Torfaen 5 Winchester 5 North Lincolnshire 5 Hambleton 5 East Staffordshire 5 Plymouth 5 Maldon 5 Sefton 5 St. Helens 5 South Northamptonshire 5 Wigan 5 Sedgemoor 5 The Vale of Glamorgan 5 Eastbourne 4 East Lindsey 4 Preston 4 Adur 4 Kettering 4 Exeter 4 Monmouthshire 4 South Norfolk 4 Test Valley 4 East Dorset 4 South Hams 4 Shropshire 4 Eastleigh 4 Newcastle-under-Lyme 4 Newcastle upon Tyne 4 Broxtowe 4 Ashfield 4 Solihull 4 Tewkesbury 4 Richmondshire 4 Conwy 4 Doncaster 4 North Tyneside 3 Calderdale 3 Carmarthenshire 3 Telford and Wrekin 3 Derbyshire Dales 3 Stratford-on-Avon 3 North Devon 3 Blaby 3 Northumberland 3 North Somerset 3 Boston 3 Chesterfield 3 Blackburn with Darwen 3 Caerphilly 3 Denbighshire 3 Flintshire 3 Mansfield 3 Trafford 3 Ceredigion 3 East Riding of Yorkshire 3 North Warwickshire 3 Bury 3 East Hampshire 3 Lancaster 3 Knowsley 3 Isle of Wight 3 Wellingborough 3 West Dorset 2 Christchurch 2 Mid Suffolk 2 Fenland 2 Lincoln 2 Wyre Forest 2 Harrogate 2 Gedling 2 Copeland 2 Redcar and Cleveland 2 Warrington 2 Selby 2 Powys 2 Gwynedd 2 North West Leicestershire 2 East Devon 2 Bassetlaw 2 East Cambridgeshire 2 Weymouth and Portland 2 Blaenau Gwent 2 Pembrokeshire 2 Amber Valley 2 Fareham 2 Havant 2 Babergh 2 Allerdale 2 Blackpo

### Q4: How does crime type look like in London?
* Top 3 crime types are: violence/sexual offence, anti-social behaviour and vehicle crime.
* Out of 14 crime types, 6 are related to robbery/theft, i.e: other theft, burglary, theft from persion, shopliftting, robbery, bicycle theft

In [0]:
%sql
select crimeType, count(*) as number_of_crimes
from londonCrime
group by 1
order by number_of_crimes desc

crimeType number_of_crimes Violence and sexual offences 436370 Anti-social behaviour 307168 Vehicle crime 228442 Other theft 209245 Burglary 141941 Criminal damage and arson 103532 Public order 99907 Drugs 94239 Theft from the person 85154 Shoplifting 78204 Robbery 64289 Bicycle theft 41787 Other crime 20057 Possession of weapons 12248

### Q5: What does the resolution (LastOutcomeCategory) look like?

* Almost all crimes realted to drugs were dealt with. 47% were locally resolved.
* 60% of crimes related to possession of weapons were dealt with (under investigation or offenders were sentenced)
* For crime types apart from possession of weapons and drugs, only less than 20% were dealt with (under investigation or offenders were sentenced). 30% - 50% of the ivestigations were complete and no suspects were identified.

In [0]:
df2 = df.groupBy('CrimeType', 'LastOutcomeCategory').count() \
  .select('CrimeType', 'LastOutcomeCategory', col('count').alias('countByCrimeTypeAndOutcome')) \
  .withColumn('countByCrimeType', sum('countByCrimeTypeAndOutcome').over(Window.partitionBy('CrimeType')))\
  .withColumn('percentage', col('countByCrimeTypeAndOutcome')/col('countByCrimeType')*100)

In [0]:
df3 = df2.filter((df2.LastOutcomeCategory != 'Investigation complete; no suspect identified') \
                 & (df2.LastOutcomeCategory != 'Status update unavailable') \
                 & (df2.LastOutcomeCategory.isNotNull())).orderBy('percentage', ascending=False)

display(df3)

CrimeType LastOutcomeCategory countByCrimeTypeAndOutcome countByCrimeType percentage Drugs Local resolution 44313 94239 47.02193359437176 Possession of weapons Awaiting court outcome 2490 12248 20.32984977139125 Other crime Under investigation 3634 20057 18.11836266640076 Possession of weapons Court result unavailable 1808 12248 14.761593729588505 Violence and sexual offences Under investigation 62382 436370 14.295666521529895 Public order Under investigation 12716 99907 12.727836888306124 Possession of weapons Under investigation 1545 12248 12.61430437622469 Drugs Offender given penalty notice 10709 94239 11.363660480268255 Robbery Under investigation 6733 64289 10.473020267852977 Drugs Awaiting court outcome 9604 94239 10.191109837752947 Criminal damage and arson Under investigation 8461 103532 8.172352509369084 Drugs Under investigation 7265 94239 7.70912255011195 Drugs Court result unavailable 6871 94239 7.291036619658528 Shoplifting Under investigation 5433 78204 6.94721497621605 Burglary Under investigation 9610 141941 6.770418695091622 Other crime Awaiting court outcome 1267 20057 6.316996559804557 Bicycle theft Under investigation 2495 41787 5.970756455356929 Vehicle crime Under investigation 13468 228442 5.89558837691843 Other theft Under investigation 12058 209245 5.762622762790031 Other crime Local resolution 1152 20057 5.743630652639976 Theft from the person Under investigation 4766 85154 5.596918524085774 Other crime Court result unavailable 1066 20057 5.314852669890811 Possession of weapons Offender given a caution 633 12248 5.168190725016329 Shoplifting Local resolution 3555 78204 4.545803283719503 Drugs Offender given a caution 4165 94239 4.419613960250002 Shoplifting Awaiting court outcome 2899 78204 3.706971510408675 Shoplifting Court result unavailable 2685 78204 3.433328218505447 Public order Awaiting court outcome 3127 99907 3.1299108170598657 Public order Court result unavailable 2561 99907 2.563383947070776 Violence and sexual offences Awaiting court outcome 11018 436370 2.5249215115612897 Robbery Awaiting court outcome 1540 64289 2.395433122307082 Criminal damage and arson Awaiting court outcome 2443 103532 2.3596569176679676 Violence and sexual offences Court result unavailable 8711 436370 1.996241721474895 Robbery Court result unavailable 1282 64289 1.9941203005179737 Possession of weapons Local resolution 223 12248 1.8207054212932723 Burglary Awaiting court outcome 2578 141941 1.8162475958320712 Burglary Court result unavailable 2492 141941 1.755659041432708 Criminal damage and arson Court result unavailable 1748 103532 1.6883668817370474 Criminal damage and arson Local resolution 1524 103532 1.4720086543290962 Possession of weapons Offender given community sentence 159 12248 1.298171129980405 Other crime Offender given a caution 248 20057 1.2364760432766615 Criminal damage and arson Offender given a caution 1249 103532 1.206390294788085 Possession of weapons Offender sent to prison 142 12248 1.1593729588504247 Violence and sexual offences Offender given a caution 4688 436370 1.0743176662007012 Drugs Offender given a drugs possession warning 996 94239 1.0568872759685481 Possession of weapons Offender given suspended prison sentence 101 12248 0.8246244284781188 Drugs Offender fined 760 94239 0.8064601704177675 Violence and sexual offences Local resolution 3510 436370 0.8043632697023169 Shoplifting Offender given a caution 582 78204 0.7442074574190579 Public order Local resolution 736 99907 0.7366851171589578 Public order Offender given a caution 631 99907 0.6315873762599217 Public order Offender given penalty notice 591 99907 0.5915501416317175 Shoplifting Offender given penalty notice 447 78204 0.5715820162651527 Vehicle crime Awaiting court outcome 1259 228442 0.5511245742901918 Possession of weapons Defendant found not guilty 62 12248 0.5062050947093403 Possession of weapons Court case unable to proceed 62 12248 0.5062050947093403 Bicycle theft Awaiting court outcome 203 41787 0.4857970182114054 Ve

In [0]:
df4 = df2.filter(df2.LastOutcomeCategory == 'Investigation complete; no suspect identified').orderBy('percentage', ascending=False)
display(df4)

CrimeType LastOutcomeCategory countByCrimeTypeAndOutcome countByCrimeType percentage Bicycle theft Investigation complete; no suspect identified 20852 41787 49.90068681647402 Violence and sexual offences Investigation complete; no suspect identified 212259 436370 48.641978137818825 Public order Investigation complete; no suspect identified 48244 99907 48.28890868507712 Robbery Investigation complete; no suspect identified 29292 64289 45.563004557544836 Criminal damage and arson Investigation complete; no suspect identified 44997 103532 43.46192481551597 Shoplifting Investigation complete; no suspect identified 33182 78204 42.43005472865838 Other theft Investigation complete; no suspect identified 87921 209245 42.018208320389974 Vehicle crime Investigation complete; no suspect identified 92121 228442 40.3257719683771 Burglary Investigation complete; no suspect identified 54684 141941 38.525866381102006 Other crime Investigation complete; no suspect identified 7234 20057 36.06720845590068 Theft from the person Investigation complete; no suspect identified 28131 85154 33.03544167038542 Possession of weapons Investigation complete; no suspect identified 2570 12248 20.98301763553233 Drugs Investigation complete; no suspect identified 5054 94239 5.362960133278154